module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,688 @@
+
+	if isInvalidBlobError(err) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data is valid, and not sign the receipt
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("HALTING REQUESTER: invalid execution data found")
+
+		e.metrics.Halted()
+		return err
+	}
+
+	// Some or all of the blob was missing or corrupt. retry
+	if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+		logger.Error().Err(err).Msg("failed to get execution data for block")
+
+		return err
+	}
+
+	// Any other error is unexpected
+	if err != nil {
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("unexpected error fetching execution data")
+
+		ctx.Throw(err)
+	}
+
+	logger.Debug().Msg("Fetched execution data")
+
+	e.executionDataCache.Add(request.Height, executionData)
+
+	return nil
+}
+
+// fetchExecutionData fetches the ExecutionData by its ID, and times out if fetchTimeout is exceeded
+func (e *executionDataRequester) fetchExecutionData(signalerCtx irrecoverable.SignalerContext, executionDataID flow.Identifier, height uint64) (executionData *state_synchronization.ExecutionData, err error) {
+	ctx, cancel := context.WithTimeout(signalerCtx, e.config.FetchTimeout)
+	defer cancel()
+
+	start := time.Now()
+	e.metrics.ExecutionDataFetchStarted()
+	defer func() {
+		// needs to be run inside a closure so the variables are resolved when the defer is executed
+		e.metrics.ExecutionDataFetchFinished(time.Since(start), err == nil, height)
+	}()
+
+	// Get the data from the network
+	// this is a blocking call, won't be unblocked until either hitting error (including timeout) or
+	// the data is received
+	executionData, err = e.eds.Get(ctx, executionDataID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Write it to the local blobstore
+	_, _, err = e.eds.Add(signalerCtx, executionData)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to write execution data to blobstore: %w", err)
+	}
+
+	return executionData, nil
+}
+
+// Notification Worker Methods
+
+func (e *executionDataRequester) processNotificationJob(ctx irrecoverable.SignalerContext, job module.Job, complete func()) {
+	// convert job into a block entry
+	entry, err := jobs.JobToBlockEntry(job)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to convert job to entry: %w", err))
+	}
+
+	e.processNotification(ctx, entry)
+	complete()
+}
+
+func (e *executionDataRequester) processNotification(ctx irrecoverable.SignalerContext, entry *jobs.BlockEntry) {
+	e.log.Debug().Msgf("notifying for block %d", entry.Height)
+
+	// send notifications
+	e.notifyConsumers(entry.ExecutionData)
+
+	e.metrics.NotificationSent(entry.Height)
+	e.executionDataCache.Remove(entry.Height)
```

Comment:
```suggestion
	// since request has notified all consumer for this height, the cache for height lower than this 
	// won't be needed any more. because the notification is delivered by a single worker who 
	// guarantees blocks are delivered in consecutive height order.
	e.executionDataCache.Remove(entry.Height)
```

Added some comments to explain why it's the request to notify cache about the lowest height of retain cached items.

I have two comments here:
1. If it's requester to notify ExecutionDataCache what height to remove, then why `ExecutionDataCache` still needs `e.blockConsumer.LastProcessedIndex`. I thought it was so that data cache can know what items to remove by itself. I don't think we need to pass `e.blockConsumer.LastProcessedIndex` to execution data cache.
```golang
	e.executionDataCache = jobs.NewExecutionDataReader(
		e.config.MaxCachedEntries,          // max number of ExecutionData objects to cache
		e.blockConsumer.LastProcessedIndex, // method to get the highest consecutive height to notify
		e.getExecutionData,                 // method to get execution data from the db
	)
```

2. we need to be careful about memory leak risk here.
The executionDataCache relies on the caller to guarantee `Remove(height)` is called with all heights that is called by `Add`. If not, then some added items will never be removed. And when cache is full of un-removable items, then no more item will be cached. In this case, the cache will be useless. 
Ideally, instead of `Remove` a height, we could `PruneUpToHeight`, which in most cases is equivalent to removing one height. But this could guarantee items below this height will be removed and no un-removable items. There are lots of implementation in the codebase you can refer to for implementing `PruneUpToHeight`.

Reply:
1. This definitely need some more comments to clarify. `e.blockConsumer.LastProcessedIndex` tells the execution data reader what's the highest consecutive height that we have downloaded execution data for. e.g. if we have data for 1, 2, 3, 5, 6, 7, it will return 3. Which means the notifier can send notifications for heights up to 3. It's not really for the cache, though it is used to determine the max height to cache as a simplification

2. good call. I'll make that change

Reply:
> though it is used to determine the max height to cache as a simplification

Why it's the max height to cache? 

Let's walk through this case:

Blocks up to height 7 has been sealed. Now the block consumer workers have downloaded execution data for block `1, 2, 3, 5, 6`. The worker for block 4 is still downloading it. When the workers finishes `5` and `6`, they save the execution data to storage, and also to the cache. If the max height to cache is the last processed index which is `3`, then `5` and `6` will not be cached, which would cause a cache miss later. Let's see. Now the worker finished downloading `4`, it also saves to both storage and the cache, and then update the `processedNotificationIndex` to `4`. And notifier will continue working on the `5` and `6`. Since `5` and `6` are downloaded already, the notifier will read `5` and `6`, but they would have to read them from storage instead of cache, because they were not cached after downloaded.

Reply:
I've removed the caching from this PR. we can add it later